---
title: "NK cells - differential gene expression"
author: "Hildur"
date: "January, 2019"
output: word_document
---

hNK - NK cells from healthy donor mice

teNK - NK cells from tumor-bearing mice

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, cache=TRUE, message = F)
suppressMessages(library(DESeq2))
suppressMessages(library(limma))
suppressMessages(library(data.table))
suppressMessages(library(Heatplus))
suppressMessages(library(cetcolor))

outFolder = "results/"
inFolder = "data/"
name_suffix ="_final"
py_path = "/usr/bin/python"
```

## Alignment

The raw data was aligned using hisat2, converted to sorted bam files using samtools - UMI's where used to select best read quality using NuDup - and finally the counts were calculated using HTseq-count.

## Differential Expression

Differential expression was estimated using DESeq2.


Differential expression between the healthy NK and NK cells from tumor bearing mice was estimated using DESeq2 [@Love2014] by reading in the counts generated by HTSeq&nbsp;[@Anders2015].

```{r}
samplesTable <- data.frame(samplenames=c("hNK_1","hNK_2","hNK_3","hNK_4","teNK_1","teNK_2","teNK_3","teNK_4"),
                  samplefiles=c("1A_S9_grcm38","2A_S10_grcm38","3A_S11_grcm38","4A_S12_grcm38","5A_S1_grcm38","6A_S3_grcm38","7A_S5_grcm38","8A_S7_grcm38"),
                  type=c("healthy","healthy","healthy","healthy","teNK","teNK","teNK","teNK"))

samplesTable$type <- relevel(x = samplesTable$type, ref = "healthy")
```

```{r, echo=TRUE}
dir_count = paste0(inFolder,"counts/")
dds <- DESeqDataSetFromHTSeqCount(samplesTable, directory = dir_count, design = ~ type)
```
The samples are annotated with the particular mouse of origin/type (hNK or teNK) as
```{r}
knitr::kable(samplesTable)
```

DESeq2 estimates the sample sizes (sequencing depth per sample), dispersion (variance in read counts for genes for the negative binomial model), automatically to compute the differential expression fold changes and p-values.
The contrast argument states that we calculate fold change as teNK/healthy.
Alpha determines the cutoff for FRD calculation.

```{r, echo=TRUE}
dds <- DESeq(dds)
results.deg <- results(dds, tidy = TRUE, contrast = c("type", "teNK", "healthy"), alpha = 0.05)
table(results.deg$padj<0.05)
```


```{r}
results.deg$Min.reads = rowMin(assay(dds))
results.deg$neg.reads = rowSums(assay(dds)[,c("hNK_1","hNK_2","hNK_3","hNK_4")])
results.deg$pos.reads = rowSums(assay(dds)[,c("teNK_1","teNK_2","teNK_3","teNK_4")])
```
We save the differentially expressed genes. Here we use approximate FWER of 1e-6 and abs(L2FC) > 1.

```{r write-certain-genes}
MyList <- results.deg

wb = openxlsx::createWorkbook(creator = "Hildur")
openxlsx::addWorksheet(wb, sheetName = "Genes")
#openxlsx::writeDataTable(wb = wb, sheet = 1, MyList)
openxlsx::writeData(wb = wb, sheet = 1, MyList)
openxlsx::saveWorkbook(wb, file = paste0(outFolder,"GenesDE.xlsx"), overwrite = T)

SigGenes1 <- subset(MyList, padj <= 1e-06)
SigGenes <- subset(SigGenes1, log2FoldChange > 1 | log2FoldChange < -1)

wb = openxlsx::createWorkbook(creator = "Hildur")
openxlsx::addWorksheet(wb, sheetName = "Significant Genes")
#openxlsx::writeDataTable(wb = wb, sheet = 1, MyList)
openxlsx::writeData(wb = wb, sheet = 1, SigGenes)
openxlsx::saveWorkbook(wb, file = paste0(outFolder,"DE_SignificantGenes.xlsx"), overwrite = T)
```

At the ~FWER of 1e-6 we have `r dim(SigGenes1)[1]` differentially expressed genes, out of which `r dim(subset(SigGenes1, log2FoldChange >0))[1]` are upregulated and `r dim(subset(SigGenes1, log2FoldChange <=0))[1]` are downregulated.

#Summary data table
```{r put-all-results-together}
allresults = data.frame(geneid = results.deg$row, 
                         pvalue = results.deg$pvalue,
                         padj = results.deg$padj,
                         l2fc = results.deg$log2FoldChange,
                         total.hc.reads = rowSums(counts(dds)[, grep(pattern = "hNK", colnames(counts(dds)))]),
                         total.tc.reads = rowSums(counts(dds)[, grep(pattern = "teNK", colnames(counts(dds)))]),
                         stringsAsFactors = F
                         )
allresults <- as.data.table(allresults)
allresults <-allresults[(total.hc.reads+total.tc.reads)>10,]
```

## Multiple testing correction and filtering un-mapped and low read count genes

We manually set the p-values to 1 for all the genes for which DESeq2 did not calculate p-values and where the total read count is below 10. 

```{r get-genes-to-filter}
badGenesPVal = allresults[is.na(pvalue), geneid]
badGenesFDR = union(badGenesPVal, allresults[(total.hc.reads + total.tc.reads) < 10, geneid])
goodGenesFDR = setdiff(allresults$geneid, badGenesFDR)
```


Out of the total `r dim(dds)[1]` genes, `r length(badGenesFDR)` did not have a p-value or a low read count, and `r sum(! is.na(allresults$pvalue))` had a valid p-value. 

For multiple testing correction, we calculated the false discovery rate for all genes not having outlier values and with total read counts greater than or equal to 10. This limited the effective number of tests (genes) used for FDR calculations to `r length(goodGenesFDR)`.

```{r calculate-fdr}
allresults <-allresults[is.na(pvalue), pvalue := 1]
setkey(allresults,geneid)
allresults <-allresults[goodGenesFDR, FDR.pos := p.adjust(pvalue, method = "fdr")]
allresults <-allresults[badGenesFDR, FDR.pos := 1]
```


# Gene set enrichment

We downloaded the Msigdb gene sets to test for enrichment of differentially expressed genes. The mouse ensembl gene ids were converted to Human HGNC symbols to map the genes to Msigdb gene sets.

```{r human-mouse-conversion}
library(biomaRt)
human = useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl")
mouse = useMart("ENSEMBL_MART_ENSEMBL", dataset = "mmusculus_gene_ensembl")
mousegenes <- results.deg$row
hggenes = getLDS(attributes = c("ensembl_gene_id","external_gene_name"),
                  filters = "ensembl_gene_id", values = mousegenes,
                  mart = mouse, attributesL = c("hgnc_symbol"), 
                  martL = human, uniqueRows=T)
hggenes <- as.data.table(hggenes)
setkey(hggenes, HGNC.symbol)
```

Z-scores were calculated for each gene based on the p-value of differentially expression. z-scores were defined as abs(qnorm(pval/2)) * sign(log2FoldChange). For all genes with zero read counts or those that DESeq2 filtered out and a p-value of fold change estimated was not calculated, the z-score is zero.

```{r calculate-z-score}
allresults[, z.pos := abs(qnorm(pvalue/2)) * sign(l2fc)]
allresults <- allresults[is.na(z.pos), z.pos := 0]
allresults <- allresults[is.infinite(z.pos), z.pos := 100*sign(l2fc)]
```

```{r add-human-symbols}
setkey(allresults, geneid)
setkey(hggenes, Gene.stable.ID)
allresults <- merge(x = allresults, y = hggenes, by.x="geneid", by.y="Gene.stable.ID", all.x=TRUE)
allresults <- allresults[order(allresults$pvalue, decreasing = F),]
openxlsx::write.xlsx(allresults, file = paste0(outFolder,"allresults_genename",name_suffix,".xlsx"), 
                     sheetName = "DE Gene-level results", overwrite=TRUE,
                     asTable = T)
```


```{r write-z-scores}
allSigResults <- subset(allresults, pvalue <= 1e-06)
allSigResults <- subset(allSigResults, abs(l2fc) >1)
write.table(x = allSigResults[! is.na(HGNC.symbol), .(HGNC.symbol, z.pos)],
            file = "pos.zscores.txt",
            row.names = F,
            col.names = c("gene", "zscore"),
            quote = F,
            sep = "\t")
```

The enrichment of differentially expressed genes in a gene set was calculated as the p-value of the t-test comparing the absolute value of the z-score between the in-set and out of set genes. Therefore, a gene set with genes that are highly regulated, but in both directions, will still be considered significant.


```{r define-gsea}
run_gsea <- function(zscorefile, lsetfile = "c2", outputfile, dataprefix="", setprefix="", setsuffix=".all.v6.2.symbols.gmt")
{
  pypath <- function()
  {
    return(py_path)
  }
  if (! missing(outputfile)) outfile=paste0(dataprefix, outputfile)
  else outfile=paste0(dataprefix, zscorefile, "_",lsetfile,"_enriched.txt")
  
  comm = paste0(pypath()," do_gsea.py ", dataprefix, zscorefile, " ", setprefix,lsetfile,setsuffix ," ",
                outfile)
  system(command = comm)
}
```

The gene set enrichment was carried out for four gene sets for Msigdb (c2, c5, c6, h).

```{r do-gsea}

run_gsea(zscorefile = "pos.zscores.txt", lsetfile = "data/c2", outputfile = "pos_c2.txt",
         dataprefix = "")
run_gsea(zscorefile = "pos.zscores.txt", lsetfile = "data/c5", outputfile = "pos_c5.txt",
         dataprefix = "")
run_gsea(zscorefile = "pos.zscores.txt", lsetfile = "data/h", outputfile = "pos_h.txt",
         dataprefix = "")
run_gsea(zscorefile = "pos.zscores.txt", lsetfile = "data/c6", outputfile = "pos_c6.txt",
         dataprefix = "")
run_gsea(zscorefile = "pos.zscores.txt", lsetfile = "data/c7", outputfile = "pos_c7.txt",
         dataprefix = "")
```


```{r write-all-gsea}
library(openxlsx)
gsea_stats <- data.frame(DEMethod=NA, Design=NA, Collection=NA, "Number of sets"=NA, "Number pval < 0.05"=NA, "Number FDR < 0.05"=NA, check.names = F)

read_gsea <- function(setname, setdesc) {
  read_gsea_file <- function(filename) {
    set_results <- read.table(filename, sep = "\t", header = T, stringsAsFactors = F)
    set_results$padj <- p.adjust(set_results$pvalue, method = "fdr")
    set_results <- as.data.table(set_results)
    setkey(set_results, geneset)
    set_results
  }
  
  addsuffix <- function(df, suffix, exclude) {
    cn <- colnames(df)
    cn2 <- sapply(X = cn, FUN = function(cname) {
      if (cname %in% exclude) return(cname)
      paste0(cname, suffix)
    })
    colnames(df) <- cn2
    df
  }
  
  pos_set <- read_gsea_file(sprintf("pos_%s.txt",setname))
  gsea_stats <<- rbind(gsea_stats, list("DESeq2", "pos",paste0(setdesc, " (",setname,")"), dim(pos_set)[1], sum(pos_set$pvalue < 0.05),  sum(pos_set$padj < 0.05)))
  pos_set <- addsuffix(pos_set, ".pos.deseq2", c("geneset","comment","nset"))
  

  g_set_all <- pos_set

  return(g_set_all)
}



wb <- createWorkbook(creator = "Hildur")

collate_gsea_set <- function(setnameshort, setdescription, workbook) {
  g_set <- read_gsea(setname = setnameshort, setdesc = setdescription)
  shname = paste0(setnameshort," (",setdescription,")")
  addWorksheet(wb = workbook, sheetName = shname)
  # writeDataTable(wb = workbook, sheet = shname, x = g_set, colNames = T, rowNames = F, bandedRows = T, tableStyle = "TableStyleMedium2")
  writeDataTable(wb = workbook, sheet = shname, x = g_set[order(g_set$pvalue),], colNames = T, rowNames = F, bandedRows = T, tableStyle = "TableStyleMedium2")
  setColWidths(wb = workbook, sheet = shname, cols = c(1:ncol(g_set)), widths = 30, hidden = (! grepl(pattern = "(geneset)|(comment)|(padj)|(pval)|(nset)",x = colnames(g_set))))
}

addWorksheet(wb = wb, sheetName = "GSEA summary")

collate_gsea_set("c2","Curated Gene Sets",wb)
collate_gsea_set("c5","GO Gene Sets",wb)
collate_gsea_set("c6","Oncogenic Signatures",wb)
collate_gsea_set("h","Hallmarks",wb)
collate_gsea_set("c7","Immunologic signatuares",wb)


gsea_stats <- gsea_stats[]
writeDataTable(wb = wb, sheet = "GSEA summary", x = gsea_stats, rowNames = F, colNames = T, bandedRows = T, tableStyle = "TableStyleMedium2")

saveWorkbook(wb = wb, file = paste0(outFolder,"All pos GSEA",name_suffix,".xlsx"),overwrite = T)

```

Finally save the R data for further analysis and visualization of the data
```{r}
#save.image("NK2020_deg.rda")
```

